### DISCUSSION TITLE
Unable to build: "sh: tsc: command not found"

### DISCUSSION THREAD

**Discussion by brendan-jefferis** [vercel/vercel#6709] [Help]
Project used to build, just started running issues today.

This first occurred when I clicked Redeploy (without build cache) in the Vercel UI.

Since then I have not been able to run a successful build.

```
16:31:09.657  	Cloning github.com/vicinity/vcx-corporate-website (Branch: development, Commit: f131881)
16:31:10.681  	Cloning completed: 1.024s
16:31:10.710  	Analyzing source code...
16:31:11.743  	Installing build runtime...
16:31:13.574  	Build runtime installed: 1.831s
16:31:16.103  	No Build Cache available
16:31:17.258  	Installing dependencies...
16:31:17.265  	Detected `package-lock.json` generated by npm 7...
16:31:37.561  	added 1212 packages in 20s
16:31:37.562  	15 packages are looking for funding
16:31:37.562  	  run `npm fund` for details
16:31:37.582  	Detected Next.js version: 11.1.2
16:31:37.593  	Running "npm run build"
16:31:37.846  	> vcx-corporate-website@0.1.0 prebuild
16:31:37.846  	> npm run check
16:31:38.102  	> vcx-corporate-website@0.1.0 check
16:31:38.102  	> tsc && npm run test
16:31:38.107  	sh: tsc: command not found
16:31:38.125  	Error: Command "npm run build" exited with 127
```

typescript is a devDependency – shouldn't Vercel install this and add to path for every build?

**Comment by leerob** [comment]
Is `NODE_ENV` is set to `production`?

**Reply by brendan-jefferis** [reply]
Hi Lee thanks for the prompt response.

`NODE_ENV` was not explicitly set in env vars, is that normally how it's done?

After setting `NODE_ENV` to `production` for preview, development, production env vars in Vercel I get the same error result as above.

I also tried deploying from the command line with `vercel -e NODE_ENV=production` and got the same result.

-------------------------
### Removing package-lock
When I remove `package-lock.json` and redeploy I get different errors about a missing dependencies of dependencies (example below)

```
...
16:10:06.684 | Failed to compile.
16:10:06.685 | ModuleNotFoundError: Module not found: Error: Can't resolve 'moment' in '/vercel/path0/node_modules/react-moment/dist'
...
```

I also saw missing `axios` and then what looked like an error from within a css parsing library (again, missing a dependant function).

-------------------------

I feel like I want to start from scratch with this Vercel project, i.e clean any caches, rebuild, reset, etc. It seems to have gotten into a bad state

**Reply by leerob** [reply]
No, you don't need to explicitly set it, which is why I was checking 😄 

> I feel like I want to start from scratch with this Vercel project, i.e clean any caches, rebuild, reset, etc. It seems to have gotten into a bad state

It does seem like you're in a strange dependency state. It's hard for me to debug further without access to the code. Could you try starting from scratch and slowly adding back your code to see if you can pinpoint where the issue is?

**Reply by brendan-jefferis** [reply]
We isolated the issue and finally have working builds again by **downgrading to npm 6** for this project.

Evidence points to a **breaking change in Vercel 2 days ago where package-lock files using npm 7 (lockfileVersion: 2) causes incorrect package installation.**

Clean new Vercel project failed with the same error

I went back through my repo checking out old commits and deploying until I found one that worked. The only difference between that commit and the next (failing) one is the change in `package-lock.json` to `lockfileVersion: 2` – caused by me upgrading npm on my machine to v7 around that time.

It is important to note that all these commits used to work regardless of npm version up until 2 days ago.

After downgrading to npm 7, regenerating my `package-lock.json` from the latest commit of my development branch and redeploying it I got passing builds again.

Attempting the exact same build with an npm 7 generated `package-lock.json` (lockfileVersion: 2) fails with original error.

------------------------
Regarding the other dependency-of-a-dependency issues when deleting `package-lock.json` – this is because in lieu of a package-lock/lockfileVersion, next appears to use a version of npm earlier than 7 to resolve dependencies, which does not magically find packages that _should_ be included by package authors but are not.

There actually _were_ two npm packages in my solution that required deps that were a) not listed in the packages deps, and b) not installed as top-level deps in my code. npm 7 appears to resolve these, whereas npm 6 and below does not. (for reference, it was `react-moment` which needed `moment` but did not include it so if I use npm 5-6 I need to install `moment` directly, if I use npm 7 it gets handled automatically which is why I never encountered that error before.)
